### PR TITLE
Including database feature

### PR DESCRIPTION
**Pull request by Prashastmykaarma**
include the database feature in such a way that the existing functionality of Daksha remains as it is and if someone wants to save the test data in Daksha then he can set it up as well.

### CODE REVIEW

File: engine/models.py [hunk 1]
```diff
@@ -35,3 +36,12 @@ def __init__(self, name, test_status, step, failure_reason):
         self.test_status = test_status
         self.failed_step = step
         self.failure_cause = failure_reason
+
+
+class ResultStored(models.Model):
```
Comment (hriddhighosh4) [inline engine/models.py:41]: makes this TestResults

File: engine/executor.py [hunk 2]
```diff
@@ -64,7 +65,21 @@ def execute_test(test_executor: TestExecutor, email):
             if execution_result is False:
                 break
         logger.info("Test " + name + " finished, generating  result ")
-        generate_result(test_executor.test_id, execution_result, name, step, error_stack)
+        generate_result(test_executor.test_id, execution_result, name, step, error_stack )
+        
+        if(settings.TEST_RESULT_DB != None and settings.TEST_RESULT_DB == "postgres"):
```
Comment (hriddhighosh4) [inline engine/executor.py:70]: Add equalsIgnoreCase

File: engine/executor.py [hunk 3]
```diff
@@ -64,7 +65,21 @@ def execute_test(test_executor: TestExecutor, email):
             if execution_result is False:
                 break
         logger.info("Test " + name + " finished, generating  result ")
-        generate_result(test_executor.test_id, execution_result, name, step, error_stack)
+        generate_result(test_executor.test_id, execution_result, name, step, error_stack )
+
```
Comment (hriddhighosh4) [inline engine/executor.py:69]: make a method save_result_ib_db

File: engine/test_result_initializer.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+#This function initializes the test result of each YAML file in the database
+def initialize_test_result(test_id,test_yml):
+    time_now = datetime.now()
+    processed_time_now = time_now.__str__().replace(":", "_")
+    particular_test_result=models.ResultStored()
+    if(settings.TEST_RESULT_DB != None and settings.TEST_RESULT_DB == "postgres"):
+        logger.info(test_id)
+        logger.info(test_yml["name"])
+        logger.info("Initializing the entry in database of the abovestated test")
```
Comment (hriddhighosh4) [inline engine/test_result_initializer.py:14]: only test_result

File: engine/models.py [hunk 5]
```diff
@@ -21,12 +21,13 @@
...
 # Create your models here.
 class TestExecutor:
-    def __init__(self, index, test_id, variable_dict, test_yml, web_driver):
+    def __init__(self, index, test_id, variable_dict, test_yml, web_driver , initialized_test_result):
         self.index = index
         self.test_id = test_id
         self.variable_dictionary = copy.deepcopy(variable_dict)
         self.test_yml = test_yml
         self.web_driver = web_driver
+        self.initialized_test_result=initialized_test_result
```
Comment (hriddhighosh4) [inline engine/models.py:30]: self.test_result

File: engine/models.py [hunk 6]
```diff
@@ -21,12 +21,13 @@
 
 # Create your models here.
 class TestExecutor:
-    def __init__(self, index, test_id, variable_dict, test_yml, web_driver):
+    def __init__(self, index, test_id, variable_dict, test_yml, web_driver , initialized_test_result):
```
Comment (hriddhighosh4) [inline engine/models.py:24]: test_result

File: README.md [hunk 7]
```diff
@@ -34,6 +34,23 @@ Download Chromedriver according to your chrome version from https://chromedriver
 ### Remote Webdriver
 Setup selenium grid using https://www.selenium.dev/downloads/. Or if you want to use the dockerized version you can download and run selenium grid images from dockerhub: https://hub.docker.com/u/selenium.
 
+## Database
+ - The user can opt for this functionality if he/she wants to save the test results in the database. 
+ - To enable this, please provide the following environment variables:-
+    TEST_RESULT_DB , PG_DB , PG_USER , PG_PASSWORD , PG_HOST , PG_PORT
+
+ - We only support Postgresql database at this point. To enable this one has to set the environment        variable TEST_RESULT_DB to "postgres".
```
Comment (hriddhighosh4) [inline README.md:42]: why is there extra space?

File: README.md [hunk 8]
```diff
@@ -34,6 +34,23 @@ Download Chromedriver according to your chrome version from https://chromedriver
...
+ - This functionality is optional and Daksha workflow does not depend on it.
+
+### Deploying with Docker
```
Comment (hriddhighosh4) [inline README.md:45]: Deploying  Daksha with DB enabled by Docker

**Review by hriddhighosh4 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,6 +1,9 @@
 # `DAKSHA` Change Log
 
-*version*: `2.1.3`
+*version*: `2.2.0`
+
+## v 2.2.0
+1. Added the database feature in such a way that the existing functionality of Daksha remains as it
```
Comment (hriddhighosh4) [inline CHANGELOG.md:6]: 1. Added Database for Test Results in Daksha.
2. Now users can opt for test results to be saved in a Postgres database

**Review by hriddhighosh4 — CHANGES_REQUESTED** [review]

File: engine/test_result_utils.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+from .logs import *
+from daksha import settings
+from engine import models
+
+def initialize_test_result(test_id,test_yml):
+    test_result=models.TestResults()
+    if(settings.TEST_RESULT_DB != None and settings.TEST_RESULT_DB.lower() == "postgres"):
+        test_name=test_yml["name"]
+        test_result.TestUUId=test_id
```
Comment (hriddhighosh4) [inline engine/test_result_utils.py:26]: UUID

**Review by AlphaRahul16 — DISMISSED** [review]

File: README.md [hunk 11]
```diff
@@ -34,6 +34,23 @@ Download Chromedriver according to your chrome version from https://chromedriver
 ### Remote Webdriver
 Setup selenium grid using https://www.selenium.dev/downloads/. Or if you want to use the dockerized version you can download and run selenium grid images from dockerhub: https://hub.docker.com/u/selenium.
 
+## Database
+ - The user can opt for this functionality if he/she wants to save the test results in the database. 
+ - To enable this, please provide the following environment variables:-
+ TEST_RESULT_DB, PG_DB, PG_USER, PG_PASSWORD, PG_HOST, PG_PORT
```
Comment (Rishi16royy) [inline README.md]: add code format like `TEST_RESULT_DB `

Comment (Rishi16royy) [reply README.md]: text should be enclosed by ` . Do this in all the places where env variables are mentioned

File: README.md [hunk 12]
```diff
@@ -34,6 +34,23 @@ Download Chromedriver according to your chrome version from https://chromedriver
 ### Remote Webdriver
 Setup selenium grid using https://www.selenium.dev/downloads/. Or if you want to use the dockerized version you can download and run selenium grid images from dockerhub: https://hub.docker.com/u/selenium.
 
+## Database
+ - The user can opt for this functionality if he/she wants to save the test results in the database. 
+ - To enable this, please provide the following environment variables:-
+ `TEST_RESULT_DB`, PG_DB, PG_USER, PG_PASSWORD, PG_HOST, PG_PORT
```
Comment (Rishi16royy) [inline README.md]: env variables should be enclosed by ` .

File: README.md [hunk 13]
```diff
@@ -87,6 +104,25 @@ You can configure the application in a lot of ways by setting the following envi
 * **ALLOWED_HOSTS**
   * Provide a comma separated list of hosts which you'd like to be added to ALLOWED_HOSTS.
 
+* **TEST_RESULT_DB**
+  * If you want to use the Postgresql database to save your test reports, create this environment variable and set its value to postgres.
```
Comment (Rishi16royy) [inline README.md:108]: set its value to postgres --> set its value to `postgres`

File: README.md [hunk 14]
```diff
@@ -87,6 +104,25 @@ You can configure the application in a lot of ways by setting the following envi
...
+  * If you set this value as postgres, you need to provide additional environment variables.
+
+*  **PG_DB**
+  * Name of the database.If this value is not provided, the default name of database will be postgres.
```
Comment (Rishi16royy) [inline README.md:112]: Space after full stop

Comment (Rishi16royy) [reply README.md:112]: fix this in other places.

File: docker-compose-db.yml [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+  engine-apache:
+    image: httpd:2.4-alpine
+    container_name: engine-apache
+    restart: always
+    volumes:
+      - /daksha/test-data:/usr/local/apache2/htdocs/
+    ports:
+      - "8485:80"
+
+#If you want to provide external Postgresql database then please run docker-compose-external-db.yml
```
Comment (Rishi16royy) [inline docker-compose-db.yml:37]: change this to --> `#If you want to provide an external Postgres SQL database then remove the below lines .`

and then remove docker-compose-external-db.yml file

File: engine/testreport_generator.py [hunk 16]
```diff
@@ -99,7 +99,7 @@ def generate_report(test_id):
         test_result = json.dumps(test_result)
         replacement = {"test_id": test_id, "test_result": test_result, "passed_count": passed_count.__str__(),
```
Comment (Rishi16royy) [inline engine/testreport_generator.py:100]: As we are using TestUUID, we should replace TestID with TestUUID everywhere.

**Review by Rishi16royy — CHANGES_REQUESTED** [review]

**Review by Rishi16royy — DISMISSED** [review]

**Review by hriddhighosh4 — APPROVED** [review]

**Pull request merged**